File: solutions/cpp/grade-school/1/grade_school.cpp
#include "grade_school.h"
#include <algorithm>

namespace grade_school {

void school::add(const std::string& name, int grade_level) {
    _roster[grade_level].push_back(name);
    std::sort(_roster[grade_level].begin(), _roster[grade_level].end());
}

const school::NameVector& school::grade(int grade_level) const {
    if (_roster.find(grade_level) == _roster.end()) {
        static const NameVector empty;
        return empty;
    }
    return _roster.at(grade_level);
}

}  // namespace grade_school

File: solutions/python/grade-school/1/grade_school.py
class School(dict):
    def __init__(self):
        self.added_status = []

    def add_student(self, name, grade):
        if name in self:
            self.added_status.append(False)
        else:
            self[name] = grade
            self.added_status.append(True)

    def roster(self):
        def inverted(item):
            a, b = item
            return b, a
        return [
            name for (name, grade) in 
            sorted(self.items(), key = inverted)
        ]

    def grade(self, grade_number):
        return sorted([
            name for (name, grade) in self.items()
            if grade == grade_number
        ])

    def added(self):
        return self.added_status
